### PR TITLE
AwsSignatureHandler now overrides the synchronous Send() method in .NET 5.

### LINKED ISSUE #478: HttpClient.Send (synchronous API) is not intercepted
**Describe the bug**

The package correctly intercepts `HttpClient.SendAsync()` (and all the extension methods that invoke it, such as `GetAsync()` or `PostAsync()`), adding the Authorization header and such.

However, the synchronous API is not properly intercepted. `HttpClient.Send()` does not cause the necessary headers to be added.

Presumably, this is because this API was added in .NET 5. A new target would be needed.

**To Reproduce**

In .NET 5, create an `HttpClient` configured with this package's message handler, which normally signs any request made. Construct an `HttpRequestMessage` and send it using `HttpClient.Send()`. After sending, confirm the absence of headers on the `HttpRequestMessage`. Compare this to `HttpClient.SendAsync()`, which causes the headers to be added as expected.

**Expected behavior**

On send, the necessary headers are expected to be added to the `HttpRequestMessage`, such as the Authorization header.

**Environment**

.NET 5.

**Additional context**

It would probably make sense to include .NET 5 as an additional target platform, so that the synchrous API can be ~~intercepted~~ overridden.

There are significant use cases for the HttpClient's synchronous API. The [original proposal](https://github.com/dotnet/runtime/issues/32125) contains extensive explanations and discussions about it. It was finally added in .NET 5.

Stephen Toub [explains](https://github.com/dotnet/runtime/issues/32125#issuecomment-585361666):

> The key is [sync-over-async] ends up requiring many more threads in order to sustain throughput. And it can take a long time for the thread pool to ramp up to the necessary level, while in the interim the system can essentially be deadlocked. That means either the app-level dev needs to explicitly set a high min thread count to force the ramp up, or we need to make the pool way more aggressive at thread injection, which harms other cases.

My own experience confirms that thread pool exhaustion happens _extremely_ quickly with sync-over-async - much more so than one might expect. This limits the number of simultaneous operations to a fraction of what is achievable with a purely synchronous API.

In a .NET 5 preview blog, Stephen Toub further [summarizes](https://devblogs.microsoft.com/dotnet/performance-improvements-in-net-5/) the addition of a synchronous API:

> While HttpClient was designed for asynchronous usage, we have found situations where developers are unable to utilize asynchrony, such as when implementing an interface method that’s only synchronous, or being called from a native operation that requires a response synchronously, yet the need to download data is ubiquitous. In these cases, forcing the developer to perform “sync over async” (meaning performing an asynchronous operation and then blocking waiting for it to complete) performs and scales worse than if a synchronous operation were used in the first place. As such, .NET 5 sees limited new synchronous surface area added to HttpClient and its supporting types. dotnet/runtime does itself have use for this in a few places. For example, on Linux when the X509Certificates support needs to download a certificate as part of chain building, it is generally on a code path that needs to be synchronous all the way back to an OpenSSL callback; previously this would use HttpClient.GetByteArrayAsync and then block waiting for it to complete, but that was shown to cause noticeable scalability problems for some users…

### PR DESCRIPTION
**Pull request by Timovzl**
fix: Synchronous HttpClient.Send().

AwsSignatureHandler now overrides the synchronous Send() method in .NET 5.

Closes #478.

_I did not have time to conform to the contribution guidelines to the letter. Feel free to use this PR as you see fit. If you prefer to reject it and copy relevant changes to your own PR, I'm all for it._

# Description

Please include a summary of the change and which issue is fixed. Please also include relevant motivation and context.

Fixes # (issue)

# Checklist

- [ ] I have squashed my commits into a single one with a message that aligns with the contributing guidelines
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works

### CODE REVIEW

File: src/Private/ContentHash.cs [hunk 1]
```diff
@@ -25,8 +24,8 @@ public static async Task<string> CalculateAsync(HttpContent? content)
                 return AWS4Signer.EmptyBodySha256;
             }
 
-            var data = await content.ReadAsByteArrayAsync();
-            var hash = AWS4Signer.ComputeHash(data);
+            var contentStream = await content.ReadAsStreamAsync();
```
Comment (Timovzl) [inline src/Private/ContentHash.cs:28]: @FantasticFiasco I thought of doing this, but decided it might be a downgrade from the existing code. `await ... ReadAsStreamAsync()` only waits until it can start streaming, I believe. When we pass the stream to be read by a sychronous method, it will be streamed synchronously. As such, we introduce the risk of doing part of the IO synchronously, which made the original async implementation seem superior to me.

Or do we know the response body to be always miniscule?

Comment (Timovzl) [reply src/Private/ContentHash.cs:28]: It's possible this is mitigated by the fact that when the stream becomes available, some number of bytes is generally available. Might be sufficient if the body is known to be tiny.

Comment (FantasticFiasco) [reply src/Private/ContentHash.cs:28]: > `await ... ReadAsStreamAsync()` only waits until it can start streaming, I believe

Didn't know that, will have to read up on it. Do you have any links regarding it?

> Or do we know the response body to be always miniscule?

No we don't. The body can be of any size, and we should support them all.

Comment (Timovzl) [reply src/Private/ContentHash.cs:28]: > Do you have any links regarding it?

Actually, it's good that you asked. It turns out that there is some [confusion](https://github.com/dotnet/runtime/issues/31316) about it:

- `HttpClient.GetStreamAsync()` definitely does _not_ buffer. (Not what we are interested in here, but definitely a cause of confusion if it turns out to behave differently from `HttpContent`.)
- `HttpContent.ReadAsStreamAsync()` _does_ buffer.
- However, "the CreateContentReadStreamAsync method it uses is virtual, with the intent that any self-respecting implementation overrides it to do something better" (Stephen Toub).

By the way, it took me a while to realize that we are merely reading our own output stream. That should get either buffered (usually) or re-streamed for us (such as with streaming file uploads).

Whenever we do have a buffered request body, as in the usual case, I believe I read somewhere that `ReadAsStreamAsync()` should actually reset its position and return that stream itself. That's good. That reduces copying.

In conclusion: I believe the current approach is ideal after all. The PR looks good to me.

**Review by FantasticFiasco — APPROVED** [review]

**Comment by FantasticFiasco** [comment]
I think I'm done with the review. Do you have anything else to add or mention?

File: src/Private/System/Web/HttpUtility.cs [hunk 2]
```diff
@@ -57,7 +57,7 @@ public override string ToString()
                     return "";
                 }
 
-                StringBuilder sb = new StringBuilder();
+                StringBuilder sb = new();
```
Comment (Timovzl) [inline src/Private/System/Web/HttpUtility.cs]: Nitpick: In some other places, you're using `var`. That still avoids the type name repetition, but it also keeps the type name and parameter list closer to each other, which seems more easily maintainable to me. For your consideration. :)

**Comment by Timovzl** [comment]
> I think I'm done with the review. Do you have anything else to add or mention?

Great work, nice improvements. I've added one nitpick, but I have no further additions. I appreciate you picking this up so fast.

File: src/Private/System/Web/HttpUtility.cs (hunk 2 above)
Comment (FantasticFiasco) [reply src/Private/System/Web/HttpUtility.cs]: I agree with you. I'll search for similar issues in the codebase, merge to master and release a new version tonight. A huge "thank you" for all the time and effort you've invested in this. If you encounter any other issues, don't be a stranger and let me know.
